src/main/java/gov/cabinetofice/gapuserservice/web/LoginController.java

```diff
@@ -11,37 +13,58 @@
+    private final CustomJwtServiceImpl customJwtService;
+    public static final String REDIRECT_URL_COOKIE = "redirectUrl";
+    public static final String THIRD_PARTY_COOKIE_NAME = "find-grants-test";
+    public static final String USER_SERVICE_COOKIE_NAME = "user-service-token";
 
     @GetMapping("/login")
-    public RedirectView login(final @CookieValue(name = COLA_COOKIE_NAME, required = false) String jwt,
+    public RedirectView login(final @CookieValue(name = USER_SERVICE_COOKIE_NAME, required = false) String jwt,
+                              final @CookieValue(name = THIRD_PARTY_COOKIE_NAME, required = false) String thirdPartyJwt,
                               final @RequestParam String redirectUrl,
                               final HttpServletResponse response) {
+        // TODO set strict, domain, http only etc
         response.addCookie(new Cookie(REDIRECT_URL_COOKIE, redirectUrl));
 
-        if (jwt == null) return redirectToThirdParty();
+        // Removes the prepended "s:"
+        final String colaJwt = thirdPartyJwt != null ? thirdPartyJwt.substring(2) : null;
+
+        final boolean isTokenValid = isTokenValid(jwt, () -> customJwtService.isTokenValid(jwt));
+        final boolean isThirdPartyTokenValid = isTokenValid(colaJwt, () -> thirdPartyJwtService.isTokenValid(colaJwt));
 
-        // TODO currently validating against the third party. When we switch to using custom JWTs, we'll want to validate against our own definition instead
-        final boolean isJwtValid = colaJwtService.isTokenValid(jwt);
-        if (!isJwtValid) return redirectToThirdParty();
+        if (!isTokenValid) {
+            if (!isThirdPartyTokenValid) {
+                return new RedirectView(authenticationProvider.getUrl());
+            }
+            final DecodedJWT decodedJWT = thirdPartyJwtService.decodeJwt(colaJwt);
+            final String generatedToken = customJwtService.generateToken(decodedJWT);
+            // TODO set strict, domain, http only etc
+            response.addCookie(new Cookie(USER_SERVICE_COOKIE_NAME, generatedToken));
+        }
 
         return new RedirectView("/redirect-after-login");
     }
 
     @GetMapping("/redirect-after-login")
-    public RedirectView redirectOnLogin(@CookieValue(REDIRECT_URL_COOKIE) String redirectUrl) {
-        // TODO we'll want to validate the third party JWT here once we use custom JWTs
+    public RedirectView redirectAfterLoggedIn(final @CookieValue(REDIRECT_URL_COOKIE) String redirectUrl) {
         return new RedirectView(redirectUrl);
     }
 
-    private RedirectView redirectToThirdParty() {
-        return new RedirectView(authenticationProvider.getUrl()); //TODO make sure COLA's redirect URL points to the domain the user service sits on
+    @GetMapping("/redirect-after-third-party-login")
+    //TODO make sure COLA's redirect URL points to here
+    public RedirectView redirectAfterThirdPartyLogin(final @CookieValue(REDIRECT_URL_COOKIE) String redirectUrl) {
```

Comment:
Need to consider a case where the user has somehow ended up at COLA without ever trying to access am authenticated resource in the first place, so providing a default value for `redirectUrl` that goes to the dashboard might be a good idea here.